### PR TITLE
applications: nrf_desktop: Add led stream to gaming mouse

### PR DESCRIPTION
**Pull request by mip1-nordic**
applications: nrf_desktop: Add led stream to gaming mouse

Create led_stream modul and change leds module to handle stream of
led_effect steps from configurator.

Jira:DESK-431

Signed-off-by: Michal Patyk <michal.patyk@nordicsemi.no>

### CODE REVIEW

**Comment by mip1-nordic** [comment]
@MarekPieta

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    }
+}
+
+static void handle_incoming_step(const struct config_event *event)
+{
+    store_data(&event->dyndata);
+    if (!first_sended){
+        LOG_INF("Sending first led effect");
+        send_data_from_queue();
+        first_sended = true;
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/led_stream.c]: /s/sended/sent

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            return false;
+        }
+    }
+
+    if (is_led_ready_event(eh)) {
+        handle_led_ready_event(cast_led_ready_event(eh));
+        return false;
+    }
+
+    if (is_module_state_event(eh)) {
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/led_stream.c]: Set MODULE_STATE_READY here - in reaction e.g. for led module being ready (we probably should not send events to `leds` before they are ready)

File: applications/nrf_desktop/src/events/led_ready_event.c [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/events/led_ready_event.c]: done

File: applications/nrf_desktop/src/events/led_ready_event.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/events/led_ready_event.h]: done

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+#define MODULE led_stream
+#include "module_state_event.h"
+
+#include "button_event.h"
+#include "config_event.h"
+#include "led_event.h"
+#include "led_ready_event.h"
+#include "ble_event.h"
+
+#include <logging/log.h>
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_LED_STREAM_LOG_LEVEL);
+
+static struct led_effect led_stream_effect = {
+        .step_count = 1,
+        .loop_forever = false,
+};
+static struct led_effect_step steps_queue[QUEUE_SIZE];
+static s8_t last_added_position = -1;
+static s8_t send_position = 0;
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+static bool first_sended = false;
+
+static void queue_data(const struct event_dyndata *dyndata, u8_t position)
+{
+    LOG_INF("Insert_data_to_effect");
+    size_t pos = 0;
+    steps_queue[position].color.c[0]=dyndata->data[pos];
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/modules/led_stream.c (hunk 1 above)
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/modules/led_stream.c (hunk 2 above)
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/events/led_ready_event.c [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/events/led_ready_event.c]: done

File: applications/nrf_desktop/src/events/led_ready_event.c [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+#include "led_ready_event.h"
+
+
+static int log_led_ready_event(const struct event_header *eh, char *buf,
+			  size_t buf_len)
+{
+	return snprintf(buf, buf_len, "leds are redy");
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/events/led_ready_event.c]: done

File: applications/nrf_desktop/src/events/config_event.h [hunk 10]
```diff
@@ -76,6 +77,9 @@ extern "C" {
 #define DFU_REBOOT	0x3
 #define DFU_IMGINFO	0x4
 
+/* Config event, DFU group macros */
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/events/config_event.h]: done

File: applications/nrf_desktop/src/modules/Kconfig.led_stream [hunk 11]
```diff
@@ -0,0 +1,17 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/Kconfig.led_stream]: done

File: applications/nrf_desktop/src/modules/Kconfig.led_stream [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+if DESKTOP_LED_ENABLE
+
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/Kconfig.led_stream]: done

**Comment by mip1-nordic** [comment]
@MarekPieta

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        last_added_position++;
+        LOG_INF("Insert data on position %u, free places %u", last_added_position, free_places);
+        queue_data(data, last_added_position);
+        free_places--;
+    } else {
+        LOG_INF("Queue is full - dropping incoming step");
+    }
+}
+
+static void send_data_from_queue()
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/led_stream.c]: add void as argument if function takes 0 arguments

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+    fetch_event->channel_id = event->channel_id;
+
+    fetch_event->dyndata.data[0] = free_places;
+
+    EVENT_SUBMIT(fetch_event);
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+    if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/led_stream.c]: WE may think about some sort of depends on here (no way to use led_stream without config channel)

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+static bool leds_initialized;
+
+static void queue_data(const struct event_dyndata *dyndata, u8_t position)
+{
+    LOG_INF("Insert_data_to_effect");
+    BUILD_ASSERT(CONFIG_DESKTOP_LED_COUNT == 3);
+    size_t pos = 0;
+    memcpy(steps_queue[position].color.c, &dyndata->data[pos], 3);
+    pos += 3;
+    steps_queue[position].substep_count=dyndata->data[pos];
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/led_stream.c]: does `queue_data` need position as argument

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    __ASSERT_NO_MSG(led_event->led_effect->steps);
+
+    EVENT_SUBMIT(led_event);
+}
+
+static void store_data(const struct event_dyndata *data)
+{
+    if (free_places > 0) {
+        if (last_added_position == QUEUE_SIZE - 1) {
+            last_added_position = -1;
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/led_stream.c]: maybe set to zero and increment after adding ? Then you could use unsigned types as indexes too

File: applications/nrf_desktop/src/hw_interface/leds.c [hunk 17]
```diff
@@ -191,12 +245,31 @@ static bool event_handler(const struct event_header *eh)
 	if (is_led_event(eh)) {
 		const struct led_event *event = cast_led_event(eh);
 
+#ifdef CONFIG_DESKTOP_LED_STREAM_ENABLE
+		if (event->stream) {
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/hw_interface/leds.c]: led->stream?

File: applications/nrf_desktop/src/hw_interface/leds.c [hunk 18]
```diff
@@ -6,36 +6,44 @@
...
-	struct led_color color;
-	const struct led_effect *effect;
-	u16_t effect_step;
-	u16_t effect_substep;
+    struct device *pwm_dev;
+
+    size_t id;
+    struct led_color color;
+    const struct led_effect *effect;
+#ifdef CONFIG_DESKTOP_LED_STREAM_ENABLE
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/hw_interface/leds.c]: discussed offline

File: applications/nrf_desktop/src/hw_interface/leds.c [hunk 19]
```diff
@@ -52,6 +60,45 @@ static void pwm_off(struct led *led)
 	pwm_out(led, &nocolor);
 }
 
+#ifdef CONFIG_DESKTOP_LED_STREAM_ENABLE
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/hw_interface/leds.c]: discussed offline

File: applications/nrf_desktop/src/hw_interface/leds.c [hunk 20]
```diff
@@ -52,6 +60,45 @@ static void pwm_off(struct led *led)
...
+
+static bool handle_led_stream(struct led *led)
+{
+    if(led->next_effect == NULL){
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/hw_interface/leds.c]: done

File: applications/nrf_desktop/src/hw_interface/leds.c [hunk 21]
```diff
@@ -52,6 +60,45 @@ static void pwm_off(struct led *led)
...
+
+    streaming = true;
+    led->effect = &led_off;
+}
+
+static void handle_next_incoming(const struct led_event *event)
+{
+    LOG_INF("Handle incoming effect");
+    if (leds[event->led_id].next_effect != NULL) {
+        LOG_ERR("Effect in use, override effect");
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/hw_interface/leds.c]: done

File: applications/nrf_desktop/src/hw_interface/leds.c (hunk 17 above)
Comment (mip1-nordic) [reply applications/nrf_desktop/src/hw_interface/leds.c]: done

File: applications/nrf_desktop/src/modules/CMakeLists.txt [hunk 22]
```diff
@@ -32,6 +32,9 @@ target_sources_ifdef(CONFIG_DESKTOP_USB_ENABLE
 target_sources_ifdef(CONFIG_DESKTOP_LED_ENABLE
 		     app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/led_state.c)
 
+target_sources_ifdef(CONFIG_DESKTOP_LED_ENABLE
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/CMakeLists.txt]: done

File: applications/nrf_desktop/src/modules/Kconfig.led_stream [hunk 23]
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+if DESKTOP_LED_ENABLE
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/Kconfig.led_stream]: done

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+static s8_t send_position;
+static u8_t free_places = QUEUE_SIZE;
+static bool first_sent;
+static bool leds_initialized;
+
+static void queue_data(const struct event_dyndata *dyndata, u8_t position)
+{
+    LOG_INF("Insert_data_to_effect");
+    BUILD_ASSERT(CONFIG_DESKTOP_LED_COUNT == 3);
+    size_t pos = 0;
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 25]
```diff
@@ -0,0 +1,205 @@
...
+    memcpy(steps_queue[position].color.c, &dyndata->data[pos], 3);
+    pos += 3;
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/modules/led_stream.c (hunk 15 above)
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/modules/led_stream.c [hunk 26]
```diff
@@ -0,0 +1,205 @@
...
+    led_event->led_effect=effect;
+
+    __ASSERT_NO_MSG(led_event->led_effect->steps);
+
+    EVENT_SUBMIT(led_event);
+}
+
+static void store_data(const struct event_dyndata *data)
+{
+    if (free_places > 0) {
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: discussed offline

File: applications/nrf_desktop/src/modules/led_stream.c (hunk 16 above)
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: discussed offline

File: applications/nrf_desktop/src/modules/led_stream.c (hunk 13 above)
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/modules/led_stream.c (hunk 14 above)
Comment (mip1-nordic) [reply applications/nrf_desktop/src/modules/led_stream.c]: done

File: applications/nrf_desktop/src/hw_interface/leds.c [hunk 27]
```diff
@@ -6,36 +6,44 @@
...
 #include "led_event.h"
+#include "led_ready_event.h"
 
 #include "leds_def.h"
 
 #define MODULE leds
 #include "module_state_event.h"
 
 #include <logging/log.h>
+
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/hw_interface/leds.c]: done

File: applications/nrf_desktop/src/hw_interface/leds.c [hunk 28]
```diff
@@ -6,36 +6,44 @@
...
 
 struct led {
-	struct device *pwm_dev;
-
-	size_t id;
-	struct led_color color;
-	const struct led_effect *effect;
-	u16_t effect_step;
-	u16_t effect_substep;
+    struct device *pwm_dev;
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/src/hw_interface/leds.c]: done

File: applications/nrf_desktop/configuration/nrf52840_pca20041/leds_def.h [hunk 29]
```diff
@@ -25,5 +25,10 @@ static const size_t led_pins[CONFIG_DESKTOP_LED_COUNT]
 		DT_NORDIC_NRF_PWM_PWM_1_CH0_PIN,
 		DT_NORDIC_NRF_PWM_PWM_1_CH1_PIN,
 		DT_NORDIC_NRF_PWM_PWM_1_CH2_PIN
-	}
+	},
+    {
```
[the remark this answers is not in the document]
Comment (mip1-nordic) [reply applications/nrf_desktop/configuration/nrf52840_pca20041/leds_def.h]: done

**Pull request closed without merging**